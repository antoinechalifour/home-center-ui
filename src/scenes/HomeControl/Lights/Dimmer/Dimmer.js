import React, { Component } from 'react'
import PropTypes from 'prop-types'
import debounce from 'debounce'
import styled from 'styled-components'
import InputRange from 'ui/InputRange'
import Row from 'ui/Row'
import Editable from 'components/Editable'

export default class Dimmer extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    bri: PropTypes.number.isRequired,
    setBrightness: PropTypes.func.isRequired,
    updateLightInformation: PropTypes.func.isRequired
  }

  constructor (props) {
    super(props)

    this.state = {
      value: this.props.bri
    }

    this.updateBrightness = debounce(this.updateBrightness.bind(this), 500)
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.bri) {
      this.setState({
        value: nextProps.bri
      })
    }
  }

  onNameChange = value => {
    this.props.updateLightInformation({
      variables: {
        input: {
          id: this.props.id,
          name: value
        }
      }
    })
  }

  onBrightnessChange = value => {
    this.setState({ value }, () => this.updateBrightness())
  }

  updateBrightness () {
    this.props.setBrightness({
      variables: {
        input: {
          id: this.props.id,
          bri: Math.round(this.state.value)
        }
      }
    })
  }

  render () {
    return (
      <Row align='center'>
        <div flex>
          <Name>
            <Editable onChange={this.onNameChange}>
              {this.props.name}
            </Editable>
          </Name>
          <InputRange
            value={this.state.value}
            onChange={this.onBrightnessChange}
          />
        </div>
      </Row>
    )
  }
}

const Name = styled.div`
  margin-bottom: 8px;
`
