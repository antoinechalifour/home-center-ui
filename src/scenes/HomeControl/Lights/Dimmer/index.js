import { graphql, compose } from 'react-apollo'
import { updateLight, toggleLight } from 'queries/lights'
import Dimmer from './Dimmer'

const updateLightOptions = {
  name: 'updateLight'
}

const toggleLightOptions = {
  name: 'toggleLight',
  options: {
    refetchQueries: ['GetLights']
  }
}

export default compose(
  graphql(updateLight, updateLightOptions),
  graphql(toggleLight, toggleLightOptions)
)(Dimmer)
