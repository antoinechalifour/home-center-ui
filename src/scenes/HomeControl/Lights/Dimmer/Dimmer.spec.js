jest.mock('debounce', () => fn => fn)

import 'jest-styled-components'
import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import { shallow } from 'enzyme'
import Dimmer from './Dimmer'

const renderer = new ShallowRenderer()

describe('scnees/HomeControl/Lights/Dimmer', () => {
  it('Should render correctly', () => {
    const props = {
      id: '2',
      name: 'My light #2',
      status: 'on',
      bri: 33,
      toggleLight: jest.fn(),
      updateLight: jest.fn()
    }
    const tree = renderer.render(<Dimmer {...props} />)

    expect(tree).toMatchSnapshot()
  })

  it('Should update the light when the input changes', () => {
    const props = {
      id: '3',
      name: 'My light #3',
      status: 'on',
      bri: 12,
      toggleLight: jest.fn(),
      updateLight: jest.fn()
    }
    const wrapper = shallow(<Dimmer {...props} />)

    wrapper.instance().onChange(33.5)

    expect(props.updateLight.mock.calls.length).toBe(1)
    expect(props.updateLight.mock.calls[0][0]).toEqual({
      variables: {
        lightId: '3',
        bri: 34
      }
    })
  })
})
