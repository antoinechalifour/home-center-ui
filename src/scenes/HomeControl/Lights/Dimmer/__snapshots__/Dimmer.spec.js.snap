// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scnees/HomeControl/Lights/Dimmer Should render correctly 1`] = `
<Flex(row)
  align="center"
  justify="flex-start"
>
  <StatusIcon
    isOn={true}
    onClick={[Function]}
  />
  <div
    flex={true}
  >
    <styled.div>
      My light #2
    </styled.div>
    <InputRange
      onChange={[Function]}
      value={33}
    />
  </div>
</Flex(row)>
`;
